imxrt-iomuxc/src/imxrt101x/mod.rs

```diff
@@ -0,0 +1,100 @@
+//! # Example
+//!
+//! In the example below, we implement a hypothetical `uart_new` function, which is responsible
+//! for preparing a UART peripheral. To properly configure the peripheral, we need the two
+//! pads that represent a peripheral's TX and RX pins. The implementation will use the
+//! `imxrt_iomuxc::uart::prepare()` function to prepare the pins.
+//!
+//! Note the trait bounds on `uart_new`. The usage requires that
+//!
+//! - the user provides one TX and one RX pin
+//! - the modules for each pin match
+//!
+//! ```no_run
+//! use imxrt_iomuxc as iomuxc;
+//! use iomuxc::uart::{Pin, TX, RX};
+//!
+//! # struct UART;
+//! /// Creates a UART peripheral from the TX and RX pads, and a baud rate
+//! fn uart_new<T, R>(mut tx: T, mut rx: R, baud: u32) -> UART
+//! where
+//!     T: Pin<Direction = TX>,
+//!     R: Pin<Direction = RX, Module = <T as Pin>::Module>,
+//! {
+//!     // Check the imxrt-iomuxc documentation to understand why
+//!     // this is unsafe.
+//!     unsafe {
+//!         iomuxc::uart::prepare(&mut tx);
+//!         iomuxc::uart::prepare(&mut rx);
+//!     }
+//!     // Prepare the rest of the UART peripheral, and return it...
+//!     # UART
+//! }
+//!
+//! # let sd_12 = unsafe { imxrt_iomuxc::imxrt101x::sd::SD_12::new() };
+//! # let sd_11 = unsafe { imxrt_iomuxc::imxrt101x::sd::SD_11::new() };
+//! // SD_12 and SD_11 are a suitable pair of UART pins
+//! uart_new(sd_12, sd_11, 115_200);
+//! ```
+//!
+//! Specifically, the trait bounds guard against non-UART pins:
+//!
+//! ```compile_fail
+//! # use imxrt_iomuxc as iomuxc;
+//! # use iomuxc::uart::{Pin, TX, RX};
+//! # struct UART;
+//! # fn uart_new<T, R>(mut tx: T, mut rx: R, baud: u32) -> UART
+//! # where
+//! #     T: Pin<Direction = TX>,
+//! #     R: Pin<Direction = RX, Module = <T as Pin>::Module>,
+//! # {
+//! #     unsafe {
+//! #         iomuxc::uart::prepare(&mut tx);
+//! #         iomuxc::uart::prepare(&mut rx);
+//! #     }
+//! #     UART
+//! # }
+//! # let ad_10 = unsafe { imxrt_iomuxc::imxrt101x::ad::AD_10::new() };
+//! # let ad_11 = unsafe { imxrt_iomuxc::imxrt101x::ad::AD_11::new() };
+//! // Neither pad is a valid UART pin
+//! uart_new(ad_10, ad_11, 115_200);
+//! ```
+//!
+//! It also guards against mismatched UART pins:
+//!
+//! ```compile_fail
+//! # use imxrt_iomuxc as iomuxc;
+//! # use iomuxc::uart::{Pin, TX, RX};
+//! # struct UART;
+//! # fn uart_new<T, R>(mut tx: T, mut rx: R, baud: u32) -> UART
+//! # where
+//! #     T: Pin<Direction = TX>,
+//! #     R: Pin<Direction = RX, Module = <T as Pin>::Module>,
+//! # {
+//! #     unsafe {
+//! #         iomuxc::uart::prepare(&mut tx);
+//! #         iomuxc::uart::prepare(&mut rx);
+//! #     }
+//! #     UART
+//! # }
+//! # let gpio_09 = unsafe { imxrt_iomuxc::imxrt101x::gpio::GPIO_09::new() };
+//! # let gpio_10 = unsafe { imxrt_iomuxc::imxrt101x::gpio::GPIO_10::new() };
+//! // GPIO_10 is a UART1 TX pin, and GPIO_09 is a UART1 RX pin
+//! uart_new(gpio_10, gpio_09, 115_200);
```

Comment:
This documentation test was designed to fail. It should show that supplying pads for separate UART peripherals does not compile.

https://github.com/imxrt-rs/imxrt-rs/blob/017a970596a59145dce29342fb2a6e45f2fa5722/imxrt-iomuxc/src/imxrt106x/mod.rs#L85-L88

It looks like these are two UART1. Consider mismatching the UART pins.

Reply:
👍